src/toil/lib/docker.py

```diff
@@ -0,0 +1,255 @@
+        defer = _docker.RM
+    if '--rm' in baseDockerCall and defer is not _docker.RM:
+        _logger.warn('--rm being passed to docker call but defer not set to dockerCall.RM, defer set to: ' + str(defer))
+        defer = _docker.RM
+    job.defer(_dockerKill, containerName, action=defer)
+    # Defer the permission fixing function which will run after this job concludes.
+    # We call this explicitly later on in this function, but we defer it as well to handle unexpected job failure.
+    job.defer(_fixPermissions, tool, workDir)
+
+    # Make subprocess call
+    call = baseDockerCall + [tool] + parameters
+    _logger.info("Calling docker with " + repr(call))
+
+    if outfile:
+        subprocess.check_call(call, stdout=outfile)
+    else:
+        if checkOutput:
+            return subprocess.check_output(call)
+        else:
+            subprocess.check_call(call)
+
+
+_docker.FORGO = 0
+_docker.STOP = 1
+_docker.RM = 2
+
+
+def _dockerKill(containerName, action):
+    """
+    Kills the specified container.
+    :param str containerName: The name of the container created by docker_call
+    :param int action: What action should be taken on the container?  See `defer=` in
+           :func:`docker_call`
+    """
+    running = _containerIsRunning(containerName)
+    if running is None:
+        # This means that the container doesn't exist.  We will see this if the container was run
+        # with --rm and has already exited before this call.
+        _logger.info('The container with name "%s" appears to have already been removed.  Nothing to '
+                  'do.', containerName)
+    else:
+        if action in (None, _docker.FORGO):
+            _logger.info('The container with name %s continues to exist as we were asked to forgo a '
+                      'post-job action on it.', containerName)
+            return
+        else:
+            _logger.info('The container with name %s exists. Running user-specified defer functions.',
+                         containerName)
+            if running and action >= _docker.STOP:
+                _logger.info('Stopping container "%s".', containerName)
+                subprocess.check_call(['docker', 'stop', containerName])
+            else:
+                _logger.info('The container "%s" was not found to be running.', containerName)
+            if action >= _docker.RM:
+                # If the container was run with --rm, then stop will most likely remove the
+                # container.  We first check if it is running then remove it.
+                running = _containerIsRunning(containerName)
+                if running is not None:
+                    _logger.info('Removing container "%s".', containerName)
+                    try:
+                        subprocess.check_call(['docker', 'rm', '-f', containerName])
+                    except subprocess.CalledProcessError:
+                        _logger.exception("'docker rm' failed.")
```

Comment:
As a user of the API, my general feeling is that I'd prefer to get a `RuntimeException` back if this failed (or at least the option to get an exception back), rather than having the exception get eaten silently. Like, I'm thinking of how we'd use these in the Spark service jobs to kill the Spark/HDFS containers (which run in a daemon-ish mode). If I try to kill those and I fail, then the nodes are in a weird state and I likely can't run a new Spark/HDFS container on those nodes (resource conflicts, IIRC). That's my general intuition.